contracts/src/v0.8/functions/dev/1_0_0/FunctionsClient.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import {Functions} from "./Functions.sol";
+import {IFunctionsRouter} from "./interfaces/IFunctionsRouter.sol";
+import {IFunctionsClient} from "./interfaces/IFunctionsClient.sol";
+import {IFunctionsCoordinator} from "./interfaces/IFunctionsCoordinator.sol";
+import {IFunctionsBilling} from "./interfaces/IFunctionsBilling.sol";
+
+/**
+ * @title The Chainlink Functions client contract
+ * @notice Contract writers can inherit this contract in order to create Chainlink Functions requests
+ */
+abstract contract FunctionsClient is IFunctionsClient {
+  IFunctionsRouter private s_router;
+  mapping(bytes32 => address) internal s_pendingRequests; /* requestId => fulfillment sender */
+
+  event RequestSent(bytes32 indexed id);
+  event RequestFulfilled(bytes32 indexed id);
+
+  error SenderIsNotRegistry();
+  error RequestIsAlreadyPending();
+  error RequestIsNotPending();
+
+  constructor(address router) {
+    setRouter(router);
+  }
+
+  /**
+   * @notice Sends a Chainlink Functions request to the stored oracle address
+   * @param req The initialized Functions.Request
+   * @param subscriptionId The subscription ID
+   * @param callbackGasLimit gas limit for the fulfillment callback
+   * @return requestId The generated request ID
+   */
+  function _sendRequest(
+    Functions.Request memory req,
+    uint64 subscriptionId,
+    uint32 callbackGasLimit,
+    bytes32 donId
+  ) internal returns (bytes32) {
+    bytes memory requestData = Functions.encodeCBOR(req);
+    bytes32 requestId = _sendRequestBytes(requestData, subscriptionId, callbackGasLimit, donId);
+    return requestId;
+  }
+
+  /**
+   * @notice Sends a Chainlink Functions request to the stored oracle address
+   * @param data The initialized Functions request data
+   * @param subscriptionId The subscription ID
+   * @param callbackGasLimit gas limit for the fulfillment callback
+   * @return requestId The generated request ID
+   */
+  function _sendRequestBytes(
+    bytes memory data,
+    uint64 subscriptionId,
+    uint32 callbackGasLimit,
+    bytes32 donId
+  ) internal returns (bytes32) {
+    bytes32 requestId = s_router.sendRequest(
+      subscriptionId,
+      data,
+      Functions.REQUEST_DATA_VERSION,
+      callbackGasLimit,
+      donId
+    );
+    s_pendingRequests[requestId] = s_router.getRoute(donId);
```

Comment:
Technically, since the request commitment is stored in the router contract, the client contract shouldn't need to spend the gas to store the pending request Id, but can rely on the security provided by the router contract, correct?

Up to you if you want to remove it, but it would save our customers a bit of gas.